### PR TITLE
Adding support for syslog+mtls drains in the syslog-agent

### PR DESCRIPTION
**Pull request by fhambrec**
# Description

This change aims to introduce mtls for syslog-drain connections by using client provided certificates and keys. Together with the mtls feature the change also addresses possible network load issues by changing the API payload of the consumed resources: The appid-centric payload was changed to a cert/key-centric payload. This loggregator-agent-release requires a new cloud controller endpoint to work. The suggested proposal and its update can be found [here](https://docs.google.com/document/d/1TkieqfYw9hrNw1xgtoPo6CYTjvEDnsIWt00ctzV57zw/edit?usp=sharing).

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Testing performed?

- [x] Unit tests
- [ ] Integration tests
- [ ] Acceptance tests

## Checklist:

- [x] This PR is being made against the `main` branch, or relevant version branch
- [x] I have made corresponding changes to the documentation
- [x] I have added testing for my changes

If you have any questions, or want to get attention for a PR or issue please reach out on the [#logging-and-metrics channel in the cloudfoundry slack](https://cloudfoundry.slack.com/archives/CUW93AF3M)

### CODE REVIEW

**Comment by Benjamintf1** [comment]
I'm pretty concerned about making this a breaking change. When deploying, it can take a pretty long time to go from "start of deploy" to "end of deploy". In the mean time, I feel like we might have downtime for syslog drains for possibly a considerable amount of time. Given, that cf operators often setup separate Diego cells in separate deployments, this amount of time could be indefinite. 

Could we not make this a breaking change for now, make a separate endpoint with the new structure, or a query parameter to get the new output, and then deprecate the older format in another major release. This would allow syslog agents to continue to get drains from the new binding cache's during the deploy.

**Comment by acrmp** [comment]
@fhambrec Does every syslog agent hold the certs and private keys for all bindings in RAM?

How do we feel about that? It might be nice if the syslog agents only held the drain credentials for the source ids currently associated with that instance but it would add complexity to the implementation.

**Comment by fhambrec** [comment]
@Benjamintf1 That's a thing we haven't thought of so far. We were planning to implement a new endpoint in the cloud controller that coexists with the old one so that schedulers would be able to query for both structures during an update. I will discuss implementing the same idea regarding the `/binding` and `/aggregate` endpoints in the schedulers with the team.

@acrmp At the moment every syslog agent will hold the certs and private keys for all bindings in RAM, yes. I am not sure if separating credentials on syslog agents based on their currently associated source ids is feasible. Thinking about this implementation there are a few points to consider:

1. Where does the syslog-agent get the source ids from?
2. Given that source drain information is not live-streamed, what happens if an app instance is moved from one diego-cell to another? --> Log loss in such scenarios must be tackled.
3. The larger a diego cell is scaled the more 'foreign' neighbours a CF user's app will have. So by design a syslog-agent will always hold credentials of multiple users and will therefore always have this security constraint.

 
From a security perspective storing the credentials on all syslog-agents is not the optimal solution but recreating it in the RAM every 15 seconds by default instead of persisting it somewhere is the best tradeoff I can think about without a huge complexity increase at the moment. Additionally, the syslog agent is not exposed via inbound connections but only creates outbound connections to the syslog drains.

**Comment by acrmp** [comment]
@fhambrec

> I will discuss implementing the same idea regarding the /binding and /aggregate endpoints in the schedulers with the team.

Thanks!

> I am not sure if separating credentials on syslog agents based on their currently associated source ids is feasible

I tend to agree.

**Comment by fhambrec** [comment]
@Benjamintf1 @acrmp There is a new commit that makes sure backwards compatibility is ensured as requested. We are currently implementing the necessary changes on the cloud controller to go the full circle.

**Comment by Benjamintf1** [comment]
Hmm, I'm a little bit confused at the recent changes, as well as what our game-plan is in terms of bringing these changes into releases, then cf-deployment, and how we're going to incorporate testing into it. 

1) I see you accounted for backwards compatibility in the binding cache by having endpoint for older syslog agents to receive drains. This is the change I was interested in/hoping to get. 

2) What seems less clear to me is the utility in the bosh property for syslog agents. binding cache's are usually deployed(to my understanding) in the order of 1/3 vms at a time, and I don't *suspect* that it's too likely to have syslog agents *ahead* of binding cache, but is this something you suspect will show up a lot? The bigger worry of mine is older syslog-agents with newer binding caches then the opposite, but I'd be curious to know what end utility you think this change will afford. It also maybe involves a weird migration process that I'm not entirely sure what your gameplan is there.

3) ATM binding cache requires capi to be in a new version to function at all. This probably means coordinating prs and making major version bumps. Were you planning on getting the capi version bumped and released and then notifying us about that so we can merge, and cut a major version? Right now I think the likely easiest/lest changes to process path to get this out is. capi merges pr for v5 endpoint without removing v3 endpoint -> capi cuts a release for v5 endpoint in addition to v3 -> cf deployment merges a pr to upgrade capi -> cf deployment ups that version of cf-deployment to canonical -> our environment pipelines pick up the new version of cf-deployment -> we approve the pr to run against pipeline and it all goes green -> merge pr -> cut major release of loggregator-agent. 

     a) If we supported both v3 and v5 internal endpoints we might be able to avoid some coordination of prs, and maybe even avoid cutting a major. We wouldn't have to wait for the pr to capi per say, and I'd feel a lot more confertable pulling in the loggregator-agent changes without the capi changes, cause everything would operateish the same by default yeah? binding caches would upgrade and use the old capi endpoints, and syslog-agents would use the new binding cache endpoint but get all the old but not new data. To get the new data, we'd have to change a property on the binding cache, and ensure that capi is on the new version line.
     b) we could avoid some degree of waiting by splintering in temporary opsfiles to upgrade capi ahead of the cf-deployment cycle. That said, I think that'd still require a major change on loggregator agent release, which is maybe less then ideal in some ways? It's maybe not worth the cruft. 
     
Honestly, I'm inclined on this path to: Keep it a major version bump, I'm willing to shove in a temporary ops file in the pipe to not have to wait for capi changes, but If we do it this way, I'd like the capi release to be a official final release. It's a little more work to manage major version bumps, but I think this feature is worth it, and I understand wanting to reduce cruft in having two possible data inputs and 2 possible data outputs in binding cache and instead only multiplex the output. 

(I say this as I actually don't remember deploy order and don't know how long we can have old capis and new binding cache's, but i don't remember what the result of that is, and downtime for the time that it takes for cc and binding cache to update is far more then downtime from upgrading every single diego cell)

4) One perhaps interesting and disappointing thing is we don't have higher level integration tests for this, as is? Cat's is the only higher order "full deployed cf test" we run, and that only tests: v1 works, v2 works, and log-cache works(through cli related tests), what we don't have is app drains work. Maybe this is something we can be ok with, with manual verification and middle level release tests, but it's something to point out. There's some understandable difficulties with the idea of running these tests in cats(parsing syslog is a pain and maintaining releases that parse syslog is annoying and setting up a syslog endpoint requires having a thing that can receive a raw tcp/tls stream which basically means setting up a tcp load balancer stuff), but the absence there feels a little bit sketchy given we're now making changes to this. I'd be willing to do manual verification for this myself, but I think we're not at that stage yet. 

5) as for the middle level of tests I referred to, I see you [made some changes](https://github.com/cloudfoundry/loggregator-agent-release/pull/119/files#diff-c981cf99d070bb80e66d87c29a63f831059717d556bba5ccf081ca263a539156R74-R78) and [some other changes](https://github.com/cloudfoundry/loggregator-agent-release/pull/119/files#diff-c981cf99d070bb80e66d87c29a63f831059717d556bba5ccf081ca263a539156R50-R56) to the mid level integration tests, but what confuses me in two ways. 
     1) I'd think we'd want to add a new data object and a new test case. Specifically, [I wouldn't modify any of the objects](https://github.com/cloudfoundry/loggregator-agent-release/pull/119/files#diff-c981cf99d070bb80e66d87c29a63f831059717d556bba5ccf081ca263a539156R74-R78) for which a test relies on currently. 
     2) I also think we'd want to add a new data object for a mut tls drain, and a new test case that see's that mutual tls actually works. Specifically, [the test server](https://github.com/cloudfoundry/loggregator-agent-release/blob/c15aff6b407d523e846aedf0f7a6f5252e755128/src/cmd/syslog-agent/app/syslog_agent_test.go#L835-L841) for that test(either just a boring copy paste or a cleaner refactor) should verify that the agent is *actually* providing a certificate and key with the right drain ca rather then only having it setup so that the agent can verify the server ca. Some sort of "When mutual tls is configured" "It can send logs to a mutual tls drain endpoint" test. I don't think you need a negative test, i think it's hard to impossible to write a negative test case here that's actually interesting/increases code quality. The key is making sure the test fails when it's not sending the cert, which is more in the process to some degree than the actual end product of the test.

**Comment by Benjamintf1** [comment]
(do we care that there's no way to set the ca? we could add this later right? or is the assumption that nobody would want to override their bosh trust store/etc for a specific drain?)

**Comment by fhambrec** [comment]
Thanks for the feedback @Benjamintf1!

I will try and answer your questions one by one, yet I think the release path and interoperability of binding-cache and agent are somehow connected.

2.: One idea of the change is that the order of VM updates does not matter:

- syslog-binding is updated first: the 'legacy' `/binding` endpoint is still offered, old syslog-agents will still receive 'legacy' data without certificates, new syslog-agents can be configured either way (mtls or tls)
- syslog-agents are updated first: the 'legacy' `binding` endpoint is still prioritised due to the default value of the `legacy_behaviour` bosh property

After a successful deployment of binding-caches and agents, cf operators can choose whether they would like to keep the tls structure or switch to the now available `/v2/bindings` endpoint to enable mtls support.

3.: Polling for `v4` (is it v3 or v4?) or `v5` depending on properties in the binding-cache is possible in general, but seemed to introduce additional data-parsing complexity from our perspective, which is likely to be removed in a soonish release as the newer endpoints offer to configure drains in either way (tls or mtls). Therefore, we decided to only query a newly introduced cc endpoint, which will be designed to focus its payload on certificate and private-key data. 
You already correctly outlined the release plan we currently think of for bringing in the change this way. I agree that having old capis and new binding-caches should definitely avoided in this scenario and we should check the deploy order first to make sure this does not happen.

Regarding 4. and 5. I agree that writing tests that come close to a fully fledged system is difficult to achieve. So far we have tried to isolate each component and create test cases for the legacy and new approach. Regarding mtls functionality, especially 5ii is something we missed. We will revisit the corresponding tests.

Regarding your additional comment: At the moment no CA configuration/rotation is supported in our change. I think we can add this later but we have not yet evaluated the complexity/implementation effort for this feature.

**Comment by dark5un** [comment]
Hi @Benjamintf1,
Thank you for your comments and we appreciate your time looking into this. Could you rephrase or analyse a bit more the comment no 5? For **5i** I might have to explain a bit more on the decision we made to change the `binding` struct and why it needs to stay changed. 

We are trying to avoid the transformation of data as much as possible especially because:
1. the previous structure was not drain centric and was inefficient due to duplication of the drains which now can carry certificates.
2. it seems easier to have and maintain only one data structure that contains all the information needed for both mtls and non-mtls drains. Unless we want mtls feature to be a completely separate part of the syslog-agent which could be ignored/removed at anytime. 
3. the `binding` structure has already changed in syslog-binding-cache and it is shared by the poller throughout all referenced code including the syslog-agent client.

Finally I would expect syslog-agent to serve both mtls and non-mtls drains simultaneously. As such, tests might need to look hybrid as well. And this is were **5ii** comes in. I agree on proving mtls data are present and a connection is feasible.

**Comment by Benjamintf1** [comment]
@dark5un I think there's a misunderstanding. 5i (changing the test data structures) isn't about the change in the data structure(the binding.Binding object itself), but the instantiated data objects that exist as text fixtures within the integration tests(links should link to specific line numbers). I would not think changing the data fixtures are adding confidence in our existing tests, and adding a new test with a new fixture that ensures mtls would be ideal instead.

**Comment by dark5un** [comment]
Hi @Benjamintf1. As you might notice in the latest commit I have reverted the aggregate payload back to the original allowing the syslog-agent test to be almost identical to the ones were before we started this feature. I know that you asked for the text fixtures to be the same. Doing that introduced a few complexities that in the end it was not worth dealing with. There are two kinds of behaviours for the agent now which have separate tests. I will slightly enrich the mtls one with some more testing on the mtls specifics shortly.
I hope we have not troubled you too much with this. Thank you again for your help. Please let us know if there is something else we could look at.

**Comment by Benjamintf1** [comment]
Oh shoot :/ I was mostly meaning, not adding mutual tls stuff to the old tests. Although was it also reverted to make the "works with old api" work?

**Comment by Benjamintf1** [comment]
btw, what's the roadmap/next steps for this?

**Comment by dark5un** [comment]
> Oh shoot :/ I was mostly meaning, not adding mutual tls stuff to the old tests. Although was it also reverted to make the "works with old api" work?

No worries. We have made sure that we are backwards compatible which should make things easier to check. On top of that we have a separate test file which includes the "legacy" tests under mtls and some more. There is a new tls server that supports client authentication. Apart from the happy path, we are testing for the following unhappy ones:
- not accepting untrusted drain certificates
- not accepting empty certificates 
- not activating drains with invalid certs
Do you see any other possible scenarios we need to cover for this feature?

**Comment by Benjamintf1** [comment]
https://github.com/cloudfoundry/cloud_controller_ng/pull/2903 cross link here for refernece

**Comment by Benjamintf1** [comment]
1) While creating syslog drains with mutual tls could be a vector for performance issues, we do have at least one method of protection against it(you can assign max number of service instances in a space as a protection measure), and it's not something we've heard complaints from operators about(despite large environments with many syslog drains). Trying to fit protections into the cf api rather than systems like how syslog agent only sends to 5 drains is almost certainly the right idea, but I see no reason to gate *this pr* on that behavior.

2) yup, makes sense

3) yup, makes sense. Could be introduced now or later. 

4) Side-note, my understanding is that cert expiration/updating would be handled by the developer issuing a cf update-user-provided-service command. I don't think there exists a easier way to notify developers/setup easy certificate rotation, but it's something we should think about. (maybe send messages over the wire that say the cert is about to expire?). 

Next steps wise, I think the capi team is still working on some unrelated pipeline problems/etc, but my expectation is still that we'd wait until our capi changes go through, then look into getting this pr merged. I was less worried about broader roadmap changes, more concerned about the path to getting the other pr merged into capi, this merged into loggregator-agent, and both into cf-deployment to go out to users. 

~ Ben & @acrmp

**Comment by Benjamintf1** [comment]
I think that should have merged in the linting pr changes. It looks like it's now complaining about linting errors introduced in this pr. I believe the merge is probably clean/fine, didn't have to deal with much conflicts.

**Comment by ctlong** [comment]
@Benjamintf1 the merge commit now present in this PR looks kinda wonky. It isn't going to affect the git history if it's merged in, right? That's not something I'm used to seeing from a rebase.

**Comment by stefanlay** [comment]
For documentation: The related PR adding the new syslog drain endpoint to the cloud controller https://github.com/cloudfoundry/cloud_controller_ng/pull/2903 was merged.

**Comment by Benjamintf1** [comment]
I think we're waiting on.
1) Maybe capi release getting cut? (maybe something we don't have to worry about?)
2) getting linting errors. fixed.

**Comment by fhambrec** [comment]
Hi @Benjamintf1 ,
we took care about the lint errors (which mostly were introduced by our recent changes, not by the rebase). We also decided to change the data structure slightly to support multiple certificate/key pairs per syslog-drain-url. We adapted the cloud controller accordingly and will soon open a PR in the cloud_controller_ng repo. 

The changes will be applied to the existing /v5 endpoint which is not yet used. No new PR for the loggregator should be required. Also, the data structure change is not a breaking change as long as we wait for the next cloud_controller_ng PR to be merged.

**Comment by Benjamintf1** [comment]
How does multiple certs/keys work? I thought you could only present one certificate/key for authentication in a mutual tls connection?

**Comment by fhambrec** [comment]
You're right, for one connection one can still only present one certificate/key. However, in the previous approach the assumption was that one syslog-drain-url will always use the same certificate/key. A drawback of this implementation was that multiple apps sending logs to the same syslog-server would use the same cert/key pair. We changed this to allow each single application to have its own cert/key pair. 

The structure for a binding went from this:
```
binding: {
url: someURL
cert: someCert
key: someKey
Apps: applications[...]
}

```
to this:
```
binding:
{
url: someURL
credentials: [
    {
    cert: someCert
    key: someKey
    apps: applications [...]
    }, 
    ...
]
}
```
This means for one URL multiple certs for different applications can be parsed. Since the binding.Bindings retrieved by the poller are transformed into syslog.Bindings before using them in the connections this change does not affect the syslog-agent but only the syslog-binding-cache and the data transformation.

**Comment by chombium** [comment]
Hi @Benjamintf1 ,

With the latest changes we wanted to lower the load on the cloud controller which the syslog binding cache does when polling the bindings. If two or more bindings share the same syslog drain url we are packing the multiple credentials in a single response from the binding polling requests to a single structure in order to lower down the number of requests. From a user perspective we are thinking that there is a valid scenario that multiple teams from a single company (real world organization) have:
- app(s) in a single space which are using multiple syslog drains connected to the same url. Each app has it's own credentials and it's up to the syslog server to do something per client if that's needed
- multiple spaces and they are using the same syslog drain url. In such case multiple syslog drains will use a single url and possibly multiple cert-key pairs to identify themselves depending on the cf org and space in which the apps are running.

This way, the apps can be identified separately or many apps can be identified as a single entity in case for example a whole team needs to have separate syslog drain. The way it is built now enables higher flexibility when configuring the syslog drains and covers many use cases based on how the apps are deployed in orgs and spaces and connected to the syslog drains. On the other side, the technical implementation optimizes the communication between the syslog binding connector's poller and the cloud controller's api.

I hope this makes the things a bit clearer.

**Comment by Benjamintf1** [comment]
Oh, I entirely misunderstood. I thought we were going back to not being able to share certs/keys between bindings on the wire with the diagram(I didn't read it carefully enough), and at first thought it was "multiple keys and certs for a binding". 

That all makes sense. I guess this means different urls can't share the same cert/keys on the wire, but that...wasn't a thing before either. 

I went to look for the cloud_controller pr you're referring to and I can't find it. Could you lmk when you want me to run ci/when you want us to merge? 

I assume dark5un is working on getting cla authorized/bot recognized for it? It's a absolute pain :(, but it's best practice that I pay attention to it. 

Also, I'll be out @ kubecon/cf-day next week, especially travelling on monday and thursday, so heads up on that front. 

Excited to get this merged in, and thank you so much for your help on this.

**Comment by fhambrec** [comment]
Hi @Benjamintf1 ,
The latest changes for the cloud controller can be found [here](https://github.com/cloudfoundry/cloud_controller_ng/pull/3028).  Regarding the CLA thing: dark5un is not with the company anymore. I had a short talk with him and we agreed that we change the author tag of his commits to me or chombium. 

With the new binding structure it is possible to link multiple cert/key pairs to one binding-url (resulting in multiple open connections in the syslog-agent) but also to link the same cert/key pair to multiple binding-urls if the different urls accept the pairs as credentials of course. The latter will result in redundancies in the api response but offers more flexibility overall (have a look at the linked PR's spec file for further info). 

I think as with the first cloud_controller PR we have to wait until a capi release is cut. After that we can let the ci run and merge this change.

**Comment by Benjamintf1** [comment]
`change the author tag of his commits to me or chombium` I'm willing to not worry about it tbh. I just want to make things as easy as possible while trying to do the right thing. It sounds like it might be hard to get the bot to confirm, but also we have a understanding that at the time of writing/etc those commits would have been covered by the cla, then I'm ok with going forward even if the checkmark is red.

**Comment by chombium** [comment]
@Benjamintf1,
thanks for your understanding and willingness to help. At the time when @dark5un created these commits, he was covered with SAP's CLA, and it would be really nice if he is credited for his contributions. We are reworking some parts of the changes, now when he's not with the company anymore, and that's why the CLA check is red.

**Comment by fhambrec** [comment]
I think this PR is ready to be merged from a coding perspective. This PR consumes the endpoint introduced in the latest cloud controller changes which are already released with CAPI 1.140. However, we require the cloud controller to be updated first to make the endpoint queryable at first. We either rely on a sequential update procedure in which the api gets updated first or we add a disclaimer to first update to CAPI 1.140 before continuing with this change. What's your opinion on that?

**Comment by fhambrec** [comment]
@Benjamintf1 A new capi release has been cut in the meantime. Considering the dependency of this PR to the new cloud controller endpoint, how would you like to proceed with this change?

**Comment by Benjamintf1** [comment]
Heads up, we just discovered that there *are* cats tests for drains. I think they could be better(see: make the tests use a tcp load balancer rather then echoing our container port and trying to read that off the logs and then setting drains or whatever, plus might be necessary for tls type tests), but they exist. I'm working on bringing them into pipelines.

**Comment by Benjamintf1** [comment]
Thanks for your patience. I'm still trying to get us to a place where we're ready to merge this. I think I still don't entirely understand the legacy_mode setup and it's uses, and I think it's important to know for integration work down the line. 

It looks like we have new capi in develop and not main, which means we could run some *manual verification* but can't really run against pipelines, at least for new behavior.

**Review by Benjamintf1 — APPROVED** [review]
running ci

File: jobs/loggr-syslog-agent/spec
Comment (ctlong) [reply jobs/loggr-syslog-agent/spec:165]: I guess an alternative would be to label it more like `cache.experimental_use_v2` and then remove that property down the line once we feel enough time has passed for folks to have upgrade to the new binding cache.

**Comment by Benjamintf1** [comment]
sorry, a lot of things came up, we didn't have time to look at cats today.

**Comment by Benjamintf1** [comment]
https://github.com/cloudfoundry/cf-acceptance-tests/pull/643 Something like this was what I was thinking. That help?

**Comment by ctlong** [comment]
🤔 Hey folks, I've been thinking about the the issue of trying to keep user-created syslog drains working during upgrades some and wanted to restart a conversation about it.

🗣️ **We previously discussed using separate releases and documentation to prevent log downtime**
Stage loggregator-agent releases with the feature flag turned off - and then later turned on - such that the binding caches have a chance to roll and start serving the v2 endpoint, leading to no downtime of user-created syslog drains.

The obvious risk is that operators would consume the cf-d releases / loggregator-agent releases out of order and experience downtime.

✨ **What if we introduced magic to prevent log downtime?**
We could build magic into the syslog agent and/or the binding cache such that they can work with either v4 or v5 simultaneously.

For example, we could have the binding cache switch to serving the new format on the same route as before and update the syslog agent to accept either format from the binding cache. The binding cache is (to my knowledge) only used by the syslog agent, which is in the same release, so there's no need to preserve its existing external API.

❓ **What are your thoughts?**
* Did you all experiment with something like this?
* Strong opinions either way?

**Comment by fhambrec** [comment]
Apart from this PR we experimented with the CF acceptance tests and had a look at the changes you proposed:

Adding tcp routes as a solution is straightforward and makes the tests very lean. Unfortunately our setup does not support tcp routing so we have to rely on the IP based solution instead.
The IP based solution is directly testable with mtls on our CF installations as we activated envoy-proxies that require client authentication anyway. Hence, we tried to keep the IP approach with a changed IP fetching behaviour:
Instead of printing the listener application's IPs and fetching the logs for configuring the writers we decided to parse the result of `cf ssh LISTENER_APP_NAME -c env` into the required variables.
 
Using that approach we found out that the successful creation of connections from a syslog-agent depends on where the application is placed:
- internal_ip only only works if the listener application is located on the same diego-cell as the syslog-agent (which logically means that the writer application itself is also located on that cell)
- external_ip only only works if the listener application is located on another diego-cell as the syslog-agent. This seemed odd to us as we expected the external_ip to also work in case of both applications being located on the same diego-cell.

If we connect to the envoy-proxy directly a tls handshake fails. The client certificates we used collide with the envoy-proxies when connecting to an external_ip. Setting `InsecureSkipVerify` to `true` on the syslog-agent solves this issue. Maybe this can be applied for a dev setup to run the tests?  

What do you think about combining the IP and tcp routing approaches to allow developers to configure the tests according to their running setup?

**Comment by Benjamintf1** [comment]
For the record: Me and felix talked a little bit about the fallback behavior. Felix had a really cool good idea of adding something to the api's saying they support the new behavior. I think we're waiting on that, then I can move forward on integrating.

**Comment by fhambrec** [comment]
With https://github.com/cloudfoundry/cloud_controller_ng/pull/3100 we extend the v4 response with a `v5_available` flag.

- It will always return true as soon as a cloud controller instance was successfully updated
- An updated binding-cache will not fall back to legacy mode once it receives this flag from cloud controller.
- As long as binding-cache is not updated it will ignore this flag and fall back to legacy mode
- As long as cloud controller is not updated binding-cache will fall back to legacy mode

The same logic was applied for the v2 endpoint of binding-cache, used by syslog-agent (used flag: `v2_available`.

**Comment by Benjamintf1** [comment]
we're working on this.

**Comment by Benjamintf1** [comment]
Howdy. Just wanted to throw a quick comment here before the end of the year. 

Me and Andrew were looking at this and doing some cleanup. Feel free to take a look at what we did on the 
https://github.com/cloudfoundry/loggregator-agent-release/tree/mtls-app-drains branch. We also did some testing in this pr https://github.com/cloudfoundry/cf-acceptance-tests/pull/687. My hope is to bring it up in the next working group meeting. 

I was really hoping to get this merged before the end of the year. I know this has been open for a very long time. My hope now is we can get it merged and released very early next year when we get back from vacation.

**Comment by Benjamintf1** [comment]
We did some cleanup in https://github.com/cloudfoundry/loggregator-agent-release/pull/177. Merged in #177. We'll reference this pr in the release notes as well. 

We're just waiting on the other pr to go through the pipeline to cut a release.

**Pull request closed without merging**
